### ISSUE TITLE
Symfony Role entity as a relationship throwing error

### ISSUE THREAD

**Issue by jarrodirwin** [api-platform/core#2581]
As I'm new to the API Platform I'm not sure if it's something I've done or some quirk with the way Symfony handles Roles but I've got a Role entity that is a Many-to-Many relation on another entity. 

When creating (POST) the entity with a Role relationship, I'm getting an error as at some point (possible as far up as the PropertyAccessor()->getValue()) is returning the string value of the Role and not the entity object.

It looks like ItemNormalizer() is expecting an object and not the string 'ROLE_ADMIN' or whatever role I'm trying to add as a relationship.

For other relationships the actual entity is passed and it works fine which makes me think it is something weird with how Role entities are loaded.

Also note, the entity is created and saved to the database correctly (with the role relationship) so it's just once saved and data is trying to be normalized for sending back to the browser.

I'm using JASONAPI. Here is an example POST.
```javascript
// POST /api/v2/customers-users
{
    "data": {
        "type": "CustomersUsers",
        "attributes": {
            "confirmed": true
        },
	    "relationships": {
	        "customer": {
	            "data": {
	                "type": "Customer",
	                "id": "/api/v2/customers/7"
	            }
	        },
	        "user": {
	            "data": {
	                "type": "User",
	                "id": "/api/v2/users/6"
	            }
	        },
	        "roles": {
	            "data": [{
	                "type": "Role",
	                "id": "/api/v2/roles/6"
	            }]
	        }
	    }
    }
}
```

Here's the error response:
```javascript
{
    "type": "https://tools.ietf.org/html/rfc2616#section-10",
    "title": "An error occurred",
    "detail": "Warning: get_class() expects parameter 1 to be object, string given",
    "trace": [
        {
            "namespace": "",
            "short_class": "",
            "class": "",
            "type": "",
            "function": "",
            "file": "/vagrant/vendor/api-platform/core/src/Util/ClassInfoTrait.php",
            "line": 34,
            "args": []
        },
        {
            "namespace": "ApiPlatform\\Core\\Bridge\\Symfony\\Routing",
            "short_class": "IriConverter",
            "class": "ApiPlatform\\Core\\Bridge\\Symfony\\Routing\\IriConverter",
            "type": "->",
            "function": "getObjectClass",
            "file": "/vagrant/vendor/api-platform/core/src/Bridge/Symfony/Routing/IriConverter.php",
            "line": 118,
            "args": [
                [
                    "string",
                    "ROLE_ADMIN"
                ]
            ]
        },
        {
            "namespace": "ApiPlatform\\Core\\Bridge\\Symfony\\Routing",
            "short_class": "IriConverter",
            "class": "ApiPlatform\\Core\\Bridge\\Symfony\\Routing\\IriConverter",
            "type": "->",
            "function": "getIriFromItem",
            "file": "/vagrant/vendor/api-platform/core/src/JsonApi/Serializer/ItemNormalizer.php",
            "line": 214,
            "args": [
                [
                    "string",
                    "ROLE_ADMIN"
                ]
            ]
        },
        {
            "namespace": "ApiPlatform\\Core\\JsonApi\\Serializer",
            "short_class": "ItemNormalizer",
            "class": "ApiPlatform\\Core\\JsonApi\\Serializer\\ItemNormalizer",
            "type": "->",
            "function": "normalizeRelation",
            "file": "/vagrant/vendor/api-platform/core/src/Serializer/AbstractItemNormalizer.php",
            "line": 564,
            "args": [
                [
                    "object",
                    "ApiPlatform\\Core\\Metadata\\Property\\PropertyMetadata"
                ],
                [
                    "string",
                    "ROLE_ADMIN"
                ],
                [
                    "string",
                    "App\\Entity\\Role"
                ],
                [
                    "string",
                    "jsonapi"
                ],
                [
                    "array",
                    {
                        "groups": [
                            "array",
                            [
                                [
                                    "string",
                                    "list"
                                ],
                                [
                                    "string",
                                    "confirmed"
                                ],
                                [
                                    "string",
                                    "customers/users/details"
                                ]
                            ]
                        ],
                        "operation_type": [
                            "string",
                            "collection"
                        ],
                        "collection_operation_name": [
                            "string",
                            "post"
                        ],
                        "resource_class": [
                            "string",
                            "App\\Entity\\CustomersUsers"
                        ],
                        "input": [
                            "null",
                            null
                        ],
                        "output": [
                            "null",
                            null
                        ],
                        "request_uri": [
                            "string",
                            "/api/v2/customers-users"
                        ],
                        "uri": [
                            "string",
                            "https://nzcer-assist-api.devtools/api/v2/customers-users"
                        ],
                        "resources": [
                            "object",
                            "ApiPlatform\\Core\\Serializer\\ResourceList"
                        ],
                        "resources_to_push": [
                            "object",
                            "ApiPlatform\\Core\\Serializer\\ResourceList"
                        ],
                        "cache_key": [
                            "string",
                            "530ec4bf8cf6d992926ec0b261e82b01"
                        ]
                    }
                ]
            ]
        },
        {
            "namespace": "ApiPlatform\\Core\\Serializer",
            "short_class": "AbstractItemNormalizer",
            "class": "ApiPlatform\\Core\\Serializer\\AbstractItemNormalizer",
            "type": "->",
            "function": "normalizeCollectionOfRelations",
            "file": "/vagrant/vendor/api-platform/core/src/Serializer/AbstractItemNormalizer.php",
            "line": 536,
            "args": [
                [
                    "object",
                    "ApiPlatform\\Core\\Metadata\\Property\\PropertyMetadata"
                ],
                [
                    "array",
                    [
                        [
                            "string",
                            "ROLE_ADMIN"
                        ]
                    ]
                ],
                [
                    "string",
                    "App\\Entity\\Role"
                ],
                [
                    "string",
                    "jsonapi"
                ],
                [
                    "array",
                    {
                        "groups": [
                            "array",
                            [
                                [
                                    "string",
                                    "list"
                                ],
                                [
                                    "string",
                                    "confirmed"
                                ],
                                [
                                    "string",
                                    "customers/users/details"
                                ]
                            ]
                        ],
                        "operation_type": [
                            "string",
                            "collection"
                        ],
                        "collection_operation_name": [
                            "string",
                            "post"
                        ],
                        "resource_class": [
                            "string",
                            "App\\Entity\\CustomersUsers"
                        ],
                        "input": [
                            "null",
                            null
                        ],
                        "output": [
                            "null",
                            null
                        ],
                        "request_uri": [
                            "string",
                            "/api/v2/customers-users"
                        ],
                        "uri": [
                            "string",
                            "https://nzcer-assist-api.devtools/api/v2/customers-users"
                        ],
                        "resources": [
                            "object",
                            "ApiPlatform\\Core\\Serializer\\ResourceList"
                        ],
                        "resources_to_push": [
                            "object",
                            "ApiPlatform\\Core\\Serializer\\ResourceList"
                        ],
                        "cache_key": [
                            "string",
                            "530ec4bf8cf6d992926ec0b261e82b01"
                        ]
                    }
                ]
            ]
        },
        {
            "namespace": "ApiPlatform\\Core\\Serializer",
            "short_class": "AbstractItemNormalizer",
            "class": "ApiPlatform\\Core\\Serializer\\AbstractItemNormalizer",
            "type": "->",
            "function": "getAttributeValue",
            "file": "/vagrant/vendor/api-platform/core/src/JsonApi/Serializer/ItemNormalizer.php",
            "line": 329,
            "args": [
                [
                    "object",
                    "App\\Entity\\CustomersUsers"
                ],
                [
                    "string",
                    "roles"
                ],
                [
                    "string",
                    "jsonapi"
                ],
                [
                    "array",
                    {
                        "groups": [
                            "array",
                            [
                                [
                                    "string",
                                    "list"
                                ],
                                [
                                    "string",
                                    "confirmed"
                                ],
                                [
                                    "string",
                                    "customers/users/details"
                                ]
                            ]
                        ],
                        "operation_type": [
                            "string",
                            "collection"
                        ],
                        "collection_operation_name": [
                            "string",
                            "post"
                        ],
                        "resource_class": [
                            "string",
                            "App\\Entity\\CustomersUsers"
                        ],
                        "input": [
                            "null",
                            null
                        ],
                        "output": [
                            "null",
                            null
                        ],
                        "request_uri": [
                            "string",
                            "/api/v2/customers-users"
                        ],
                        "uri": [
                            "string",
                            "https://nzcer-assist-api.devtools/api/v2/customers-users"
                        ],
                        "resources": [
                            "object",
                            "ApiPlatform\\Core\\Serializer\\ResourceList"
                        ],
                        "resources_to_push": [
                            "object",
                            "ApiPlatform\\Core\\Serializer\\ResourceList"
                        ],
                        "cache_key": [
                            "string",
                            "530ec4bf8cf6d992926ec0b261e82b01"
                        ]
                    }
                ]
            ]
        },
        {
            "namespace": "ApiPlatform\\Core\\JsonApi\\Serializer",
            "short_class": "ItemNormalizer",
            "class": "ApiPlatform\\Core\\JsonApi\\Serializer\\ItemNormalizer",
            "type": "->",
            "function": "getPopulatedRelations",
            "file": "/vagrant/vendor/api-platform/core/src/JsonApi/Serializer/ItemNormalizer.php",
            "line": 87,
            "args": [
                [
                    "object",
                    "App\\Entity\\CustomersUsers"
                ],
                [
                    "string",
                    "jsonapi"
                ],
                [
                    "array",
                    {
                        "groups": [
                            "array",
                            [
                                [
                                    "string",
                                    "list"
                                ],
                                [
                                    "string",
                                    "confirmed"
                                ],
                                [
                                    "string",
                                    "customers/users/details"
                                ]
                            ]
                        ],
                        "operation_type": [
                            "string",
                            "collection"
                        ],
                        "collection_operation_name": [
                            "string",
                            "post"
                        ],
                        "resource_class": [
                            "string",
                            "App\\Entity\\CustomersUsers"
                        ],
                        "input": [
                            "null",
                            null
                        ],
                        "output": [
                            "null",
                            null
                        ],
                        "request_uri": [
                            "string",
                            "/api/v2/customers-users"
                        ],
                        "uri": [
                            "string",
                            "https://nzcer-assist-api.devtools/api/v2/customers-users"
                        ],
                        "resources": [
                            "object",
                            "ApiPlatform\\Core\\Serializer\\ResourceList"
                        ],
                        "resources_to_push": [
                            "object",
                            "ApiPlatform\\Core\\Serializer\\ResourceList"
                        ],
                        "cache_key": [
                            "string",
                            "530ec4bf8cf6d992926ec0b261e82b01"
                        ]
                    }
                ],
                [
                    "array",
                    [
                        [
                            "array",
                            {
                                "name": [
                                    "string",
                                    "customer"
                                ],
                                "type": [
                                    "string",
                                    "Customer"
                                ],
                                "cardinality": [
                                    "string",
                                    "one"
                                ]
                            }
                        ],
                        [
                            "array",
                            {
                                "name": [
                                    "string",
                                    "user"
                                ],
                                "type": [
                                    "string",
                                    "User"
                                ],
                                "cardinality": [
                                    "string",
                                    "one"
                                ]
                            }
                        ],
                        [
                            "array",
                            {
                                "name": [
                                    "string",
                                    "roles"
                                ],
                                "type": [
                                    "string",
                                    "Role"
                                ],
                                "cardinality": [
                                    "string",
                                    "many"
                                ]
                            }
                        ]
                    ]
                ]
            ]
        },
        {
            "namespace": "ApiPlatform\\Core\\JsonApi\\Serializer",
            "short_class": "ItemNormalizer",
            "class": "ApiPlatform\\Core\\JsonApi\\Serializer\\ItemNormalizer",
            "type": "->",
            "function": "normalize",
            "file": "/vagrant/vendor/symfony/serializer/Serializer.php",
            "line": 152,
            "args": [
                [
                    "object",
                    "App\\Entity\\CustomersUsers"
                ],
                [
                    "string",
                    "jsonapi"
                ],
                [
                    "array",
                    {
                        "groups": [
                            "array",
                            [
                                [
                                    "string",
                                    "list"
                                ],
                                [
                                    "string",
                                    "confirmed"
                                ],
                                [
                                    "string",
                                    "customers/users/details"
                                ]
                            ]
                        ],
                        "operation_type": [
                            "string",
                            "collection"
                        ],
                        "collection_operation_name": [
                            "string",
                            "post"
                        ],
                        "resource_class": [
                            "string",
                            "App\\Entity\\CustomersUsers"
                        ],
                        "input": [
                            "null",
                            null
                        ],
                        "output": [
                            "null",
                            null
                        ],
                        "request_uri": [
                            "string",
                            "/api/v2/customers-users"
                        ],
                        "uri": [
                            "string",
                            "https://nzcer-assist-api.devtools/api/v2/customers-users"
                        ],
                        "resources": [
                            "object",
                            "ApiPlatform\\Core\\Serializer\\ResourceList"
                        ],
                        "resources_to_push": [
                            "object",
                            "ApiPlatform\\Core\\Serializer\\ResourceList"
                        ],
                        "cache_key": [
                            "string",
                            "530ec4bf8cf6d992926ec0b261e82b01"
                        ]
                    }
                ]
            ]
        },
        {
            "namespace": "Symfony\\Component\\Serializer",
            "short_class": "Serializer",
            "class": "Symfony\\Component\\Serializer\\Serializer",
            "type": "->",
            "function": "normalize",
            "file": "/vagrant/vendor/symfony/serializer/Serializer.php",
            "line": 125,
            "args": [
                [
                    "object",
                    "App\\Entity\\CustomersUsers"
                ],
                [
                    "string",
                    "jsonapi"
                ],
                [
                    "array",
                    {
                        "groups": [
                            "array",
                            [
                                [
                                    "string",
                                    "list"
                                ],
                                [
                                    "string",
                                    "confirmed"
                                ],
                                [
                                    "string",
                                    "customers/users/details"
                                ]
                            ]
                        ],
                        "operation_type": [
                            "string",
                            "collection"
                        ],
                        "collection_operation_name": [
                            "string",
                            "post"
                        ],
                        "resource_class": [
                            "string",
                            "App\\Entity\\CustomersUsers"
                        ],
                        "input": [
                            "null",
                            null
                        ],
                        "output": [
                            "null",
                            null
                        ],
                        "request_uri": [
                            "string",
                            "/api/v2/customers-users"
                        ],
                        "uri": [
                            "string",
                            "https://nzcer-assist-api.devtools/api/v2/customers-users"
                        ],
                        "resources": [
                            "object",
                            "ApiPlatform\\Core\\Serializer\\ResourceList"
                        ],
                        "resources_to_push": [
                            "object",
                            "ApiPlatform\\Core\\Serializer\\ResourceList"
                        ]
                    }
                ]
            ]
        },
        {
            "namespace": "Symfony\\Component\\Serializer",
            "short_class": "Serializer",
            "class": "Symfony\\Component\\Serializer\\Serializer",
            "type": "->",
            "function": "serialize",
            "file": "/vagrant/vendor/api-platform/core/src/EventListener/SerializeListener.php",
            "line": 82,
            "args": [
                [
                    "object",
                    "App\\Entity\\CustomersUsers"
                ],
                [
                    "string",
                    "jsonapi"
                ],
                [
                    "array",
                    {
                        "groups": [
                            "array",
                            [
                                [
                                    "string",
                                    "list"
                                ],
                                [
                                    "string",
                                    "confirmed"
                                ],
                                [
                                    "string",
                                    "customers/users/details"
                                ]
                            ]
                        ],
                        "operation_type": [
                            "string",
                            "collection"
                        ],
                        "collection_operation_name": [
                            "string",
                            "post"
                        ],
                        "resource_class": [
                            "string",
                            "App\\Entity\\CustomersUsers"
                        ],
                        "input": [
                            "null",
                            null
                        ],
                        "output": [
                            "null",
                            null
                        ],
                        "request_uri": [
                            "string",
                            "/api/v2/customers-users"
                        ],
                        "uri": [
                            "string",
                            "https://nzcer-assist-api.devtools/api/v2/customers-users"
                        ],
                        "resources": [
                            "object",
                            "ApiPlatform\\Core\\Serializer\\ResourceList"
                        ],
                        "resources_to_push": [
                            "object",
                            "ApiPlatform\\Core\\Serializer\\ResourceList"
                        ]
                    }
                ]
            ]
        },
        {
            "namespace": "ApiPlatform\\Core\\EventListener",
            "short_class": "SerializeListener",
            "class": "ApiPlatform\\Core\\EventListener\\SerializeListener",
            "type": "->",
            "function": "onKernelView",
            "file": "/vagrant/vendor/symfony/event-dispatcher/Debug/WrappedListener.php",
            "line": 111,
            "args": [
                [
                    "object",
                    "Symfony\\Component\\HttpKernel\\Event\\GetResponseForControllerResultEvent"
                ],
                [
                    "string",
                    "kernel.view"
                ],
                [
                    "object",
                    "Symfony\\Component\\HttpKernel\\Debug\\TraceableEventDispatcher"
                ]
            ]
        },
        {
            "namespace": "Symfony\\Component\\EventDispatcher\\Debug",
            "short_class": "WrappedListener",
            "class": "Symfony\\Component\\EventDispatcher\\Debug\\WrappedListener",
            "type": "->",
            "function": "__invoke",
            "file": "/vagrant/vendor/symfony/event-dispatcher/EventDispatcher.php",
            "line": 212,
            "args": [
                [
                    "object",
                    "Symfony\\Component\\HttpKernel\\Event\\GetResponseForControllerResultEvent"
                ],
                [
                    "string",
                    "kernel.view"
                ],
                [
                    "object",
                    "Symfony\\Component\\EventDispatcher\\EventDispatcher"
                ]
            ]
        },
        {
            "namespace": "Symfony\\Component\\EventDispatcher",
            "short_class": "EventDispatcher",
            "class": "Symfony\\Component\\EventDispatcher\\EventDispatcher",
            "type": "->",
            "function": "doDispatch",
            "file": "/vagrant/vendor/symfony/event-dispatcher/EventDispatcher.php",
            "line": 44,
            "args": [
                [
                    "array",
                    [
                        [
                            "object",
                            "Symfony\\Component\\EventDispatcher\\Debug\\WrappedListener"
                        ],
                        [
                            "object",
                            "Symfony\\Component\\EventDispatcher\\Debug\\WrappedListener"
                        ],
                        [
                            "object",
                            "Symfony\\Component\\EventDispatcher\\Debug\\WrappedListener"
                        ],
                        [
                            "object",
                            "Symfony\\Component\\EventDispatcher\\Debug\\WrappedListener"
                        ],
                        [
                            "object",
                            "Symfony\\Component\\EventDispatcher\\Debug\\WrappedListener"
                        ]
                    ]
                ],
                [
                    "string",
                    "kernel.view"
                ],
                [
                    "object",
                    "Symfony\\Component\\HttpKernel\\Event\\GetResponseForControllerResultEvent"
                ]
            ]
        },
        {
            "namespace": "Symfony\\Component\\EventDispatcher",
            "short_class": "EventDispatcher",
            "class": "Symfony\\Component\\EventDispatcher\\EventDispatcher",
            "type": "->",
            "function": "dispatch",
            "file": "/vagrant/vendor/symfony/event-dispatcher/Debug/TraceableEventDispatcher.php",
            "line": 145,
            "args": [
                [
                    "string",
                    "kernel.view"
                ],
                [
                    "object",
                    "Symfony\\Component\\HttpKernel\\Event\\GetResponseForControllerResultEvent"
                ]
            ]
        },
        {
            "namespace": "Symfony\\Component\\EventDispatcher\\Debug",
            "short_class": "TraceableEventDispatcher",
            "class": "Symfony\\Component\\EventDispatcher\\Debug\\TraceableEventDispatcher",
            "type": "->",
            "function": "dispatch",
            "file": "/vagrant/vendor/symfony/http-kernel/HttpKernel.php",
            "line": 155,
            "args": [
                [
                    "string",
                    "kernel.view"
                ],
                [
                    "object",
                    "Symfony\\Component\\HttpKernel\\Event\\GetResponseForControllerResultEvent"
                ]
            ]
        },
        {
            "namespace": "Symfony\\Component\\HttpKernel",
            "short_class": "HttpKernel",
            "class": "Symfony\\Component\\HttpKernel\\HttpKernel",
            "type": "->",
            "function": "handleRaw",
            "file": "/vagrant/vendor/symfony/http-kernel/HttpKernel.php",
            "line": 67,
            "args": [
                [
                    "object",
                    "Symfony\\Component\\HttpFoundation\\Request"
                ],
                [
                    "integer",
                    1
                ]
            ]
        },
        {
            "namespace": "Symfony\\Component\\HttpKernel",
            "short_class": "HttpKernel",
            "class": "Symfony\\Component\\HttpKernel\\HttpKernel",
            "type": "->",
            "function": "handle",
            "file": "/vagrant/vendor/symfony/http-kernel/Kernel.php",
            "line": 198,
            "args": [
                [
                    "object",
                    "Symfony\\Component\\HttpFoundation\\Request"
                ],
                [
                    "integer",
                    1
                ],
                [
                    "boolean",
                    true
                ]
            ]
        },
        {
            "namespace": "Symfony\\Component\\HttpKernel",
            "short_class": "Kernel",
            "class": "Symfony\\Component\\HttpKernel\\Kernel",
            "type": "->",
            "function": "handle",
            "file": "/vagrant/public/index.php",
            "line": 37,
            "args": [
                [
                    "object",
                    "Symfony\\Component\\HttpFoundation\\Request"
                ]
            ]
        }
    ]
}
```

Any help / advice would be great.

**Comment by jarrodirwin** [comment]
So after more investigation by trying to GET a User entity, it turns out my issue was that the getter function getRoles() on my User entity was returning an array of role names instead of the actual role entity as per the documentation at https://api.symfony.com/4.1/Symfony/Component/Security/Core/User/UserInterface.html

I've changed this to return the entity and it now normalizes correctly. JWT authentication also seems OK with this (which is why I had to change it in the first place).

Although untested, I assume if you were using the standard User entity provided by Symfony Core, you will run into this same issue whereby the result of getRoles() would be ['ROLE_USER'] and not the role entity.
